lib/Lower/IO.cpp

```diff
@@ -160,44 +154,13 @@ static mlir::FuncOp getOutputRuntimeFunc(Fortran::lower::FirOpBuilder &builder,
 static llvm::SmallVector<mlir::Value, 4>
 splitArguments(Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
                mlir::Value arg) {
-  mlir::Value zero;
-  mlir::Value one;
-  mlir::Type argTy = arg.getType();
-  mlir::MLIRContext *context = argTy.getContext();
-  const bool isComplex = argTy.isa<fir::CplxType>();
-  const bool isCharLit = isCharacterLiteral(argTy);
-  const bool isBoxChar = argTy.isa<fir::BoxCharType>();
-
-  if (isComplex || isCharLit || isBoxChar) {
-    // Only create these constants when needed and not every time
-    zero = builder.create<mlir::ConstantOp>(loc, builder.getI64IntegerAttr(0));
-    one = builder.create<mlir::ConstantOp>(loc, builder.getI64IntegerAttr(1));
-  }
-  if (isComplex) {
-    auto eleTy =
-        fir::RealType::get(context, argTy.cast<fir::CplxType>().getFKind());
-    mlir::Value realPart =
-        builder.create<fir::ExtractValueOp>(loc, eleTy, arg, zero);
-    mlir::Value imaginaryPart =
-        builder.create<fir::ExtractValueOp>(loc, eleTy, arg, one);
-    return {realPart, imaginaryPart};
-  }
-  if (isBoxChar) {
-    mlir::Type ptrTy =
-        fir::ReferenceType::get(mlir::IntegerType::get(8, context));
-    mlir::Type sizeTy = mlir::IntegerType::get(64, context);
-    mlir::Value pointerPart =
-        builder.create<fir::ExtractValueOp>(loc, ptrTy, arg, zero);
-    mlir::Value sizePart =
-        builder.create<fir::ExtractValueOp>(loc, sizeTy, arg, one);
-    return {pointerPart, sizePart};
+  if (builder.isCharacter(arg)) {
+    auto dataLen = builder.materializeCharacter(arg);
+    return {dataLen.first, dataLen.second};
   }
-  if (isCharLit) {
-    mlir::Value variable = builder.create<fir::AllocaOp>(loc, argTy);
-    builder.create<fir::StoreOp>(loc, arg, variable);
-    mlir::Value sizePart = builder.create<mlir::ConstantOp>(
-        loc, builder.getI64IntegerAttr(getLength(argTy)));
-    return {variable, sizePart};
+  if (arg.getType().isa<fir::CplxType>()) {
```

Comment:
We should consider a `builder.isComplex()` since complex comes in both an fir and mlir flavor.

Reply:
Good point, done.